lib/deimos/active_record/batch_consumption.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+module Deimos
+  module ActiveRecord
+    # Batch consume methods
+    module BatchConsumption
+      # Handle a batch of Kafka messages
+      # @param payloads [Array<Hash>] Decoded payloads.
+      # @param metadata [Hash] Information about batch, including keys.
+      def consume_batch(payloads, metadata)
+        # Create [key, value] pairs, even if no keys are decoded
+        messages = payloads.
+          zip(metadata[:keys]).
+          map { |p, k| [k, p] }
+
+        Deimos.instrument('ar_consumer.consume_batch', topic: metadata[:topic], messages: messages) do
+          slices = _slice_batch(messages)
```

Comment:
Can you add some comments as to the overall algorithm here? I'm having trouble following what the point of these slices is.

Reply:
I think a lot of the difficulty I have is in the types that are flying around here, including hashes and arrays of arrays etc. Can we bundle some of this up into a data class?